Meter.h

```diff
@@ -97,7 +98,8 @@ typedef struct MeterClass_ {
 
 typedef struct GraphData_ {
    struct timeval time;
-   double values[METER_GRAPHDATA_SIZE];
+   uint16_t numRecords;
```

Comment:
Prefer `size_t` for sizes …

Reply:
My "graph color" branch in #714 has been using this `numRecords` name. It saves me the effort of renaming the structure fields when I rebase that branch onto this one. (With colored graph, there would be more than one "values" stacked for each record, and I think "record" would be a more accurate name for describing the data.)

Reply:
@BenBE I just don't believe it can be useful to have a graph data size larger than a million; so storing it in `uint16_t` is sufficient. I can change this to `unsigned int` type if you prefer, but `size_t` would be too large IMHO.

Reply:
I don't think we need to safe on bytes here, just because a certain value is unlikely. I prefer to have proper semantics when there is an type that encapsulates "contains the number of elements in an array". That we later on filter the usable range in that variable to some "reasonable" value (like 1 million, which BTW doesn't fit in `uint16_t` :stuck_out_tongue:) is secondary. The only reason there are still larger parts of the code still using `int` for sizes instead of `size_t` is because changing the `Vector_*` and `Hashtable_*` APIs affects most of the code base and thus is preferably done in one go for a somewhat larger release …

Reply:
Considering the padding that's going to be in the layout of that struct, having it as a uint16_t isn't saving any memory as compared to a `size_t` (and there'll only be a handful of instances of this anyway, so it'd be a long way from mattering even if it did).

[I might quibble slightly with `size_t` as I'd say it's not strictly a _size_ (byte count), but an element-count, but it's more pedantic than I really care about and honestly would be fine with either an unsigned int or a size_t.]

Reply:
@zevweiss
I intend the "colored" GraphData structure to be defined like this:

```c
typedef union {
   int16_t scaleExp;
   uint16_t numDots;
   struct {
      uint8_t itemIndex;
      uint8_t details;
   } cell;
} GraphColorCell;

typedef struct GraphData_ {
   struct timeval time;
   GraphColorCell* cells;
   unsigned int numRecords;
   uint16_t recordNumCells; // or unsigned int type
} GraphData;
```

Note that there is another field after `numRecords`. I am actually more concerned on how the structure can be expanded, than the padding issue. And I'm not a fan of using `size_t` everywhere in structs (`size_t` for local variables are okay; but global variables and struct would persistently take memory, and I like to minimize memory use whenever possible).

Reply:
> And I'm not a fan of using size_t everywhere in structs (size_t for local variables are okay; but global variables and struct would persistently take memory, and I like to minimize memory use whenever possible).

Wanting to conserve memory is admirable, but it's possible to go too far.  To draw an analogy, if you're trying to save up money to buy a mansion, it doesn't make sense to try to get there by digging loose change out from the cushions of all your friends' couches -- it won't get you meaningfully closer to your goal, and your friends will be annoyed at having their living rooms ransacked.

Furthermore, in addition to the padding, and the fact that there will only be a few of these ever allocated, bear in mind that the C library's malloc implementation is probably just going to round the allocation size up to the next multiple of 16 bytes or something anyway.  There _are_ cases where saving a few bytes can matter (like for something you're allocating a large array of), but...in this case the benefit just isn't there.

Reply:
@Explorer09 Saving memory on non-repeated structures is absolutely no priority. And saving memory on repeated structures (like your graph cells) may be taken into consideration, but still has no priority unless not doing so makes that structure prohibitively large. Not the case here.

Re `size_t` is element count: I know. My comment on bytes was re the padding and other memory savings.

Furthermore: It is recommended to put element counts before an array or structure pointer. That way (especially with inline arrays) you can avoid accidentally overflowing them. Trust me, got experience with this …

Reply:
@BenBE Your reason of accidentally overflowing makes sense with inline arrays, but I'm not sure it applies to a pointer. The other reason that would convince me is that the order looks more logical when the pointer may be null, and the code is supposed to check the size before accessing the pointer.

Reply:
Yes, in the experience I referred to, the problematic array was indeed inline.

But even with dynamic (out-of-line) arrays the placement is usually with the size before the array pointer, causing the cache line to be loaded (on access of its size) before any need to access the pointer for the actual data arises (complementing your second point).